libbeat/logp/log.go

```diff
@@ -73,21 +76,50 @@ func parseSelectors(selectors []string) (map[string]struct{}, bool) {
 
 func debugMessage(calldepth int, selector, format string, v ...interface{}) {
 	if _log.level >= LOG_DEBUG && IsDebug(selector) {
-		send(calldepth+1, LOG_DEBUG, "DBG  ", format, v...)
+		send(calldepth+1, LOG_DEBUG, "DBG", format, v...)
 	}
 }
 
 func send(calldepth int, level Priority, prefix string, format string, v ...interface{}) {
+
+	message := fmt.Sprintf(format, v...)
+	timestamp := time.Now().Format(time.RFC3339)
+	var bytes []byte
+	if _log.JSON {
+		log := map[string]interface{}{
+			"timestamp": timestamp,
+			"level":     prefix,
+			"message":   message,
+		}
+		bytes, _ = json.Marshal(log)
+	} else {
+		// Creates the log message and formats it
+		bytes = []byte(fmt.Sprintf("%s %s %s", timestamp, prefix, message))
+	}
+
 	if _log.toSyslog {
-		_log.syslog[level].Output(calldepth, fmt.Sprintf(format, v...))
+		if _log.JSON {
+			_log.syslog[level].Output(calldepth, string(bytes))
+		} else {
+			_log.syslog[level].Output(calldepth, fmt.Sprintf("%s", message))
+		}
 	}
 	if _log.toStderr {
-		_log.logger.Output(calldepth, fmt.Sprintf(prefix+format, v...))
+		if _log.JSON {
+			_log.logger.Output(calldepth, string(bytes))
+		} else {
+			_log.logger.Output(calldepth, fmt.Sprintf("%s %s", prefix, message))
+		}
 	}
 	if _log.toFile {
-		// Creates a timestamp for the file log message and formats it
-		prefix = time.Now().Format(time.RFC3339) + " " + prefix
-		_log.rotator.WriteLine([]byte(fmt.Sprintf(prefix+format, v...)))
+		if _log.JSON {
+			_log.rotator.WriteLine(bytes)
+		} else {
+			// Makes sure all prefixes have the same length
+			prefix = prefix + strings.Repeat(" ", 4-len(prefix))
+			bytes = []byte(fmt.Sprintf("%s %s %s", timestamp, prefix, message))
```

Comment:
I wonder if the Sprintf, which is basically a Join, could be implemented more efficient. A benchmark between this and the `strings.Join` version would be interesting.

If it can be optimized, it's probably worth it so that we don't affect the performance in the debug logging case.